### ISSUE TITLE
Parallel File Download Support in DVCFileSystem

### ISSUE THREAD

**Issue by jakubhejhal** [iterative/dvc#10104]
I track my YOLOv5 dataset in git repository with DVC remote being Azure Blob container.

The dataset contains many small text files and the structure is the following:
```bash
.
├── images
│   ├── train
│   │   ├── 4691.jpg
│   │   ├── 4692.jpg
│   │   ├── ...
│   │   └── 5028.jpg
│   └── val
│       ├── 4753.jpg
│       ├── 4808.jpg
│       ├── ...
│       └── 5014.jpg
├── labels
│   ├── train
│   │   ├── 4691.txt
│   │   ├── 4692.txt
│   │   ├── ...
│   │   └── 5028.txt
│   └── val
│       ├── 4753.txt
│       ├── 4808.txt
│       ├── ...
│       └── 5014.txt
└── settings.yaml

7 directories, 357 files
```

In my python training pipeline I would like to recursively download the tree of files to a local machine at the beginning of the training.
I tried to use the `DVCFileSystem.get` from `dvc.api` as [in the documentation](https://dvc.org/doc/api-reference/dvcfilesystem#downloading-a-file-or-a-directory), but the files seem to be downloaded sequentially, which is slow, when the dataset contains a lot of small files.

When I fetch the files manually with multiple parallel workers, the speedup on my machine with this demo dataset is approximately 4x. 
 
Here's the code I used for the benchmark.
```python
from dvc.api import DVCFileSystem
import time
import os
import shutil
from concurrent.futures import ThreadPoolExecutor, as_completed
from fsspec import AbstractFileSystem
import tqdm


download_dir = "test_dvc_speed"
our_impl_dir = download_dir + "/our_impl"

original_impl_dir = download_dir + "/original_impl"

shutil.rmtree(download_dir, ignore_errors=True)
os.makedirs(download_dir, exist_ok=True)

repo_url = f"<github-repo_url>"
remote_path = "<path-to-remote-dir>"


def download_paralel(fs: AbstractFileSystem, dest: str, rpath: str, num_workers):
    dest = str(dest)
    os.makedirs(dest)

    if fs.isfile(rpath):
        # download single file
        fs.get_file(rpath, dest)
        return

    # download directory recursively in parallel
    remote_paths = fs.glob(os.path.join(rpath, "**"))

    files = [f for f in remote_paths if fs.isfile(f)]
    dirs = [d for d in remote_paths if fs.isdir(d)]

    # make dirs
    for d in dirs:
        os.makedirs(os.path.join(dest, os.path.relpath(d, rpath)), exist_ok=True)  # type: ignore

    # download files in parallel
    print(f"Downloading {len(files)} files to {dest}")

    with ThreadPoolExecutor(max_workers=num_workers) as executor:
        futures = []
        for rfile in files:
            file_dest = os.path.join(dest, os.path.relpath(rfile, rpath))  # type: ignore

            futures.append(executor.submit(fs.get_file, rfile, file_dest))
        for future in tqdm.tqdm(as_completed(futures), total=len(futures)):
            _ = future.result()
            pass

t = time.time()
fs = DVCFileSystem(repo_url)
download_paralel(fs, our_impl_dir, remote_path, num_workers=100)
print("Our impl:", time.time() - t)


t = time.time()
fs = DVCFileSystem(repo_url)
fs.get(rpath=remote_path, lpath=original_impl_dir, recursive=True)
print("Original impl:", time.time() - t)


###
# Our impl: 4.03s
# Original impl: 15.33s
```

I've been told by @tibor-mach, that `dvc get` uses `DVCFileSystem` behind the scenes and that `dvc get` fetches the remote files in parallel.
Is this a bug or is there another way to fetch the directory tree using `dvc` python API?

**Comment by tibor-mach** [comment]
@jakubhejhal Thanks for the issue!

cc @efiop @dberenbaum

**Comment by skshetry** [comment]
Duplicate of #8896

**Comment by skshetry** [comment]
@jakubhejhal, please check the suggestion in https://github.com/iterative/dvc/issues/8896#issuecomment-1406592386.

**Comment by skshetry** [comment]
FYI, @jakubhejhal, this was closed by #8896, and released in 3.55.1.
